File: lib/src/main/java/com/darklycoder/download/utils/DownloadTask.java
package com.darklycoder.download.utils;

import com.darklycoder.download.info.DownloadConfig;
import com.darklycoder.download.info.TaskCellInfo;
import com.darklycoder.download.interfaces.ICellTaskListener;
import com.darklycoder.download.interfaces.IProgress;

/**
 * 下载任务
 */
public class DownloadTask implements Runnable {

    private long key;
    private TaskCellInfo mInfo;
    private DownloadConfig mConfig;
    private ICellTaskListener listener;

    public DownloadTask(long key, TaskCellInfo mInfo, DownloadConfig config, ICellTaskListener listener) {
        this.key = key;
        this.mInfo = mInfo;
        this.mConfig = config;
        this.listener = listener;
    }

    @Override
    public void run() {
        boolean result = DownloadUtils.getInstance().download(mInfo, mConfig, progressListener);

        if (null == listener) {
            return;
        }

        if (result) {
            listener.onSuccess(key, mInfo);
            return;
        }

        // TODO 记录失败原因
        listener.onFail(key, mInfo, 0);
    }

    private IProgress progressListener = new IProgress() {
        @Override
        public void onProgress(float progress) {
            listener.onProgress(key, mInfo, progress);
        }
    };

}

File: lib/src/main/java/com/darklycoder/download/utils/DownloadTask.kt
package com.darklycoder.download.utils

import com.darklycoder.download.info.DownloadConfig
import com.darklycoder.download.info.TaskCellInfo
import com.darklycoder.download.interfaces.ICellTaskListener
import com.darklycoder.download.interfaces.IProgress

/**
 * 下载任务
 */
class DownloadTask(
        private val key: Long,
        private val mInfo: TaskCellInfo,
        private val mConfig: DownloadConfig,
        private val listener: ICellTaskListener?
) : Runnable {

    private val progressListener = object : IProgress {
        override fun onProgress(progress: Float) {
            listener?.onProgress(key, mInfo, progress)
        }
    }

    override fun run() {
        val result = DownloadUtils.download(mInfo, mConfig, progressListener)

        if (result) {
            listener?.onSuccess(key, mInfo)
            return
        }

        // TODO 记录失败原因
        listener?.onFail(key, mInfo, 0)
    }

}
